lib/std/smartptrs.nim

```diff
@@ -0,0 +1,195 @@
+    assert not a3.isNil
+    assert a3[] == 0
+
+  block:
+    var a1: ConstPtr[float]
+    let a2 = newConstPtr(isolate(0))
+    let a3 = a2
+
+    assert $a1 == "(nil)"
+    assert a1.isNil
+    assert $a2 == "(0)"
+    assert not a2.isNil
+    assert a2[] == 0
+    assert $a3 == "(0)"
+    assert not a3.isNil
+    assert a3[] == 0
+
+
+import std/isolation
+
+
+type
+  UniquePtr*[T] = object
+    ## Non copyable pointer to a value of type `T` with exclusive ownership.
+    val: ptr T
+
+proc `=destroy`*[T](p: var UniquePtr[T]) =
+  if p.val != nil:
+    `=destroy`(p.val[])
+    when compileOption("threads"):
+      deallocShared(p.val)
+    else:
+      dealloc(p.val)
+
+proc `=`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+  ## The copy operation is disallowed for `UniquePtr`, it
+  ## can only be moved.
+
+proc newUniquePtr*[T](val: sink Isolated[T]): UniquePtr[T] {.nodestroy.} =
+  ## Returns a unique pointer which has exclusive ownership of the value.
+  when compileOption("threads"):
+    result.val = cast[ptr T](allocShared(sizeof(T)))
+  else:
+    result.val = cast[ptr T](alloc(sizeof(T)))
+  # thanks to '.nodestroy' we don't have to use allocShared0 here.
+  # This is compiled into a copyMem operation, no need for a sink
+  # here either.
+  result.val[] = val.extract
+  # no destructor call for 'val: sink T' here either.
+
+converter convertUniquePtrToObj*[T](p: UniquePtr[T]): var T {.inline.} =
+  when compileOption("boundChecks"):
+    doAssert(p.val != nil, "deferencing nil unique pointer")
+  p.val[]
+
+proc isNil*[T](p: UniquePtr[T]): bool {.inline.} =
+  p.val == nil
+
+proc `[]`*[T](p: UniquePtr[T]): var T {.inline.} =
+  when compileOption("boundChecks"):
+    doAssert(p.val != nil, "deferencing nil unique pointer")
+  p.val[]
+
+proc `$`*[T](p: UniquePtr[T]): string {.inline.} =
+  if p.val == nil: "(nil)"
+  else: "(" & $p.val[] & ")"
+
+#------------------------------------------------------------------------------
+
+type
+  SharedPtr*[T] = object
+    ## Shared ownership reference counting pointer.
+    val: ptr tuple[value: T, atomicCounter: int]
+
+proc `=destroy`*[T](p: var SharedPtr[T]) =
+  if p.val != nil:
+    if (when compileOption("threads"):
+          atomicLoadN(addr p.val[].atomicCounter, ATOMIC_CONSUME) == 0 else:
+          p.val[].atomicCounter == 0):
+      `=destroy`(p.val[])
+      when compileOption("threads"):
+        deallocShared(p.val)
+      else:
+        dealloc(p.val)
+    else:
+      when compileOption("threads"):
+        discard atomicDec(p.val[].atomicCounter)
+      else:
+        dec(p.val[].atomicCounter)
+
+proc `=`*[T](dest: var SharedPtr[T], src: SharedPtr[T]) =
+  if src.val != nil:
+    when compileOption("threads"):
+      discard atomicInc(src.val[].atomicCounter)
+    else:
+      inc(src.val[].atomicCounter)
+  if dest.val != nil:
+    `=destroy`(dest)
+  dest.val = src.val
+
+proc newSharedPtr*[T](val: sink Isolated[T]): SharedPtr[T] {.nodestroy.} =
```

Comment:
user code needs to stay simple:
instead of `let a2 = newSharedPtr(isolate(0))`
allow: `let a2 = newSharedPtr(0)`

by doing same thing as in std/channels:
```nim
template trySend*[T](c: Channel[T], src: T): bool =
  ## Helper templates for `trySend`.
  trySend(c, isolate(src))
```

Reply:
fixed